cpp/src/rolling/detail/range_utils.cuh

```diff
@@ -0,0 +1,709 @@
+  using op     = less<T>;
+  using rev_op = greater<T>;
+};
+
+/**
+ * @brief Select the appropriate ordering comparator for the window type.
+ *
+ * @tparam Tag The type of the window.
+ * @tparam Order The sort order of the column used to define the windows.
+ */
+template <typename T, typename Tag>
+struct comparator_t {
+  order const order;
+
+  __device__ constexpr bool operator()(T const& x, T const& y) const noexcept
+  {
+    if (order == order::ASCENDING) {
+      using op = typename comparator_impl<T, Tag>::op;
+      return op{}(x, y);
+    } else {
+      using op = typename comparator_impl<T, Tag>::rev_op;
+      return op{}(x, y);
+    }
+  }
+};
+
+/**
+ * @brief Saturating addition or subtraction of values.
+ *
+ * @tparam Op The operation to perform, `cuda::std::plus` or `cuda::std::minus`.
+ *
+ * Performs `x + y` (respectively `x - y`) saturating at the numeric
+ * limits for the type, returning the value and a flag indicating
+ * whether overflow occurred.
+ *
+ * For arithmetic types, the usual arithmetic conversions are _not_
+ * applied, and so it is required that `x` and `y` have the same type.
+ *
+ * If `x` is floating, then `y` must be neither `inf` nor `nan` (not
+ * checked), otherwise behaviour is undefined. If `x` is `inf` or
+ * `nan`, then the result is the same as `x` and overflow does not
+ * occur.
+ *
+ * If `x` is a timestamp type, then `y` must be the matching
+ * `duration` type.
+ *
+ * If `x` is a decimal type, then `y` must be the matching underlying
+ * rep type, and _must_ have the same scale as `x` (not checked)
+ * otherwise behaviour is undefined.
+ *
+ * All other types are unsupported.
+ */
+template <typename Op>
+struct saturating {
+  static_assert(cuda::std::is_same_v<Op, cuda::std::plus<>> ||
+                  cuda::std::is_same_v<Op, cuda::std::minus<>>,
+                "Only for addition or subtraction");
+  template <typename T, CUDF_ENABLE_IF(cuda::std::is_floating_point_v<T>)>
+  [[nodiscard]] __host__ __device__ constexpr inline cuda::std::pair<T, bool> operator()(
+    T x, T y) noexcept
+  {
+    // Mimicking spark requirements, inf/nan x propagates
+    if (cuda::std::isinf(x) || cuda::std::isnan(x)) { return {x, false}; }
+    // Requirement, not checked, y is not inf or nan.
+    T result = Op{}(x, y);
+    // If the result is outside the range of finite values it can at
+    // this point only be +- infinity (we can't generate a nan by
+    // adding a non-nan/non-inf y to a non-nan/non-inf x).
+    if (result < cuda::std::numeric_limits<T>::lowest()) {
+      return {cuda::std::numeric_limits<T>::lowest(), true};
+    } else if (result > cuda::std::numeric_limits<T>::max()) {
+      return {cuda::std::numeric_limits<T>::max(), true};
+    }
+    return {result, false};
+  }
+
+  template <typename T, CUDF_ENABLE_IF(cuda::std::is_integral_v<T>&& cuda::std::is_signed_v<T>)>
+  [[nodiscard]] __host__ __device__ constexpr inline cuda::std::pair<T, bool> operator()(
```

Comment:
As an aside, given that `saturating` has no state, we can probably get away with `operator()` being `const`.